extmod/modtls_mbedtls.c

```diff
@@ -767,6 +771,74 @@ static MP_DEFINE_CONST_OBJ_TYPE(
 /******************************************************************************/
 // ssl module.
 
+#if MICROPY_PY_SSL_ECDSA_SIGN_ALT
+int mbedtls_ecdsa_sign_alt(mbedtls_ecp_group *grp, mbedtls_mpi *r, mbedtls_mpi *s,
+    const mbedtls_mpi *d, const unsigned char *buf, size_t blen,
+    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng) {
+
+    int ret;
+    uint8_t key[256];
+
+    if (MP_STATE_PORT(ecdsa_sign_alt_cb) == mp_const_none) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
```

Comment:
The fallback is to allow signing with a key that's Not stored on the SE (a DER on the filesystem for example). The SE can't be mutually exclusive, you may not want to use it all even if it's enabled.

The alternative is to either check if the key has magic bytes before calling the SE, or to return a special value to indicate that this is a regular key, not stored in SE.

Reply:
OK, so that's a third use case:
- alternate function provided but it decides to fall back to the default


> or to return a special value to indicate that this is a regular key, not stored in SE.

Yes, I think it should do that.  So you explicitly indicate that you want to fall back to the default.  I suggest returning `None` from the Python callback to indicate this.  Any exception that's raised by the Python callback should be reported to the user and abort the function.

Reply:
Done. If the callback returns `None` the wrapper will fall back to the default implementation. If it raises an exception, the exception is caught, printed and the proxy function returns an error to the wrapper.